### PR TITLE
Use commit SHA on PR builds

### LINKED ISSUE #41: Wrong git hash being built on PR synchronization
After pushing a new commit to a branch, Drone picks up the change but checks out the wrong git hash. Notice in the attached image that the git hash in the header (it's actually the correct one) is different from the output of `git log` in the `build` step.

screen shot 2017-03-30 at 18 31 42

### PR DESCRIPTION
**Pull request by fsouza**
Fix drone/drone#1000.
Fix #41.

This is just #38 and #50 in a single PR.

### CODE REVIEW

**Comment by bradrydzewski** [comment]
@fsouza thanks, would you mind bumping the plugin version number to `1.4`?
https://github.com/drone-plugins/drone-git/blob/master/.drone.yml#L28

**Comment by fsouza** [comment]
@bradrydzewski sure, just did! Let me know if anything doesn't look right :)

Thanks for reviewing!

File: .drone.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ pipeline:
     image: plugins/docker:17.05
     pull: true
     repo: plugins/git
-    tags: [ latest, 1.3.0, 1.3, 1 ]
+    tags: [ latest, 1.4.0, 1.4, 1.3.0, 1.3, 1 ]
```
Comment (bradrydzewski) [inline .drone.yml]: thanks, can you remove the 1.3 tags? then we are good to merge :)

```diff
-tags: [ latest, 1.4.0, 1.4, 1.3.0, 1.3, 1 ]
+tags: [ latest, 1.4.0, 1.4, 1 ]
```

Comment (fsouza) [reply .drone.yml]: Ohhhh, I see. Sorry, pushing it now.

**Comment by tonglil** [comment]
@bradrydzewski I don't agree with the change in behavior of this PR.

Effectively this changes all steps run for `event: pull_request` to the state of a repo being the push commit instead of the merge ref commit.

**Comment by bradrydzewski** [comment]
I didn't realize it wasn't using the merge commit. I will rollback until we have a better solution. I think we need a proper RFC for this because everyone seems to have a slightly different idea for how this plugin should work, and I feel like I'm in a tough spot.

**Comment by bradrydzewski** [comment]
I added a backoff to the plugin to retry when it cannot fetch the merge ref
https://github.com/drone-plugins/drone-git/commit/dd60daec4ed64250367658c9ee0d35ff73fd2ae5

@fsouza I do apologize for the confusion I've caused. I think we might need to make a change at the drone level before we can re-merge this PR. I think when using the merge ref, drone needs to capture the merge commit sha, instead of the base commit sha. Then we can re-address. Sorry, and am appreciate of the PR. We'll get it figured out :)

**Comment by geekdave** [comment]
@bradrydzewski Is it correct to assume that the backoff strategy may not prevent the ["Wrong git hash being built on PR synchronization"](https://github.com/drone-plugins/drone-git/issues/41 ) issue?  Sounds like that issue can occur when github finds the merge ref, but it's stale and refers to the previous hash in the PR.

**Comment by bradrydzewski** [comment]
@geekdave correct.

The backoff solves a problem where GitHub merge refs may not be available when at runtime, which can be an issue if GitHub or GitHub enterprise is slow to process pull requests (like it is today).

The reason I rolled back this pull request is because it was brought to my attention that this will checkout the head commit sha, and not the merge commit sha. Before we can apply this changeset we need make a change to drone core to snapshot the merge commit sha (when using the merge ref).

This is not a simple change, because the merge commit may not be available when drone tries to fetch the .drone.yml file. This means we need to implement some sort of backoff at the drone server level as well when fetching the yaml.

So there are a few different moving parts here, and I would therefore like to start with an RFC (in the core drone repository) to discuss and make sure these changes are implemented correctly, and there are no gotchas that I'm missing.

**Comment by tonglil** [comment]
Should we reopen #41 then?

**Comment by bradrydzewski** [comment]
I am not confident #41 can be implemented. GitHub does not send the merge commit sha, which means we would always checkout the unmerged, head commit.

For prior art, it seems Travis does not checkout a specific commit sha either https://travis-ci.org/urfave/cli/jobs/295356305#L431

**Pull request merged**